File: 표/5_DP/정수삼각형.py
"""
위와 같은 삼각형의 꼭대기에서 바닥까지 이어지는 경로 중, 거쳐간 숫자의 합이 가장 큰 경우 찾기
- 아래 칸으로 이동할 때는 대각선 방향으로 한 칸 오른쪽 또는 왼쪽으로만 이동 가능

*** 삼각형의 높이는 1 이상 500 이하
*** 삼각형을 이루고 있는 숫자는 0 이상 9,999 이하의 정수

각 지점마다 가질 수 있는 최댓값을 업데이트하면서 내려가면, 마지막 줄에서 답들이 모임
이들 중 최댓값만 찾으면 됨!
-> DP
"""

def solution(triangle):
    n = len(triangle)
    
    for i in range(1, n):
        length = len(triangle[i])
        for j in range(length):
            if j == 0:
                triangle[i][j] += triangle[i-1][j]
            elif j == len(triangle[i-1]):
                triangle[i][j] += triangle[i-1][j-1]
            else:
                triangle[i][j] += max(triangle[i-1][j-1], triangle[i-1][j])
    
    return max(triangle[n-1])

File: euijin/02.DP/정수삼각형.cpp
// 43분..
// DP 기본문제다.
// 아래부터 올라오면 더 편하다..(index outofbound 생각 덜 해도 됨) -> 방향도 한번 더 생각해보기


#include <string>
#include <vector>
#include <algorithm> //max있나 -> container 쓸때는 항상 iterator로 접근하는게 기본

using namespace std;

int solution(vector<vector<int>> triangle) {
    int answer = 0;
    int n = triangle.size();
    
    if(n == 1) return triangle[0][0];
    
    //전체 높이 만큼 반봅
    for(int i = 0; i < n; i++){
    //한 열 개수 만큼 반복
        for(int j = 0; j < triangle[i].size(); j++){
        
            if(i == 0) continue;// 1열이면 skip    
            //1열이 아니라면, 이전 열이랑 더하기

            //근데 기존값이랑 비교해서 최댓값 뽑아야함
            //근데 또 이전 열의 좌/우에 값이 있는지 체크해야함
            if(j == 0){//오른쪽이랑만 더하기
                triangle[i][j] = max(triangle[i][j] + triangle[i-1][j], triangle[i][j]);
                continue;
            }
            if(j == triangle[i].size() - 1){//왼쪽이랑만 더하기
                triangle[i][j] = max(triangle[i][j] + triangle[i-1][j-1], triangle[i][j]);
                continue;
            }
            //위 두개 중에 큰것 선택해서 더하기       
            triangle[i][j] = triangle[i][j] + max(triangle[i-1][j], triangle[i-1][j-1]);
        }
        
    }
    
    //마지막열 돌면서 최댓값 찾기
    //그냥 max함수로도 가능? -> iterator로
    
    // 최적화 1. vector에 max 쓰기
    // answer = max(triangle[n-1].begin(), triangle[n-1].end);
    for(int i = 0; i < triangle[n-1].size(); i++){
        answer = max(answer, triangle[n-1][i]);
    }
    
    return answer;
}